File: backend/services/call_log_service.py
import re
import time
from datetime import datetime
from typing import List, Optional
from schemas import CallLog, Merchant
from data.mock_call_logs import mock_call_logs
from services.language_detector import language_detector

class CallLogService:
    def __init__(self):
        self.call_logs: List[CallLog] = list(mock_call_logs)

    def get_call_logs(self, merchant_id: Optional[str] = None) -> List[CallLog]:
        if merchant_id:
            return [c for c in self.call_logs if c.merchantId == merchant_id]
        return self.call_logs

    def get_call_by_id(self, call_id: str) -> Optional[CallLog]:
        return next((c for c in self.call_logs if c.id == call_id), None)

    def log_call(
        self,
        conversation_id: str,
        merchant: Merchant,
        customer_transcript: str,
        agent_response: str,
        detected_intent: str,
        latency_ms: int,
        prompt_version: str
    ) -> CallLog:
        words = [w for w in agent_response.split() if w]
        word_count = len(words)
        detected_language = language_detector.detect_language(customer_transcript)

        status = 'successful'
        break_category = None
        break_reason = None
        prompt_fix = None

        if word_count > merchant.maxResponseWords + 8:
            status = 'break_detected'
            break_category = 'Long response'
            break_reason = f"Agent response ({word_count} words) exceeded merchant's maximum target limit ({merchant.maxResponseWords} words)."
            prompt_fix = f"Update system prompt rule: 'Maximum response length: {merchant.maxResponseWords} words. Be extremely brief.'"
        elif re.search(r'[\*\#\`\[\]]', agent_response) or re.search(r'^\s*\d+[\.\)]', agent_response, re.MULTILINE):
            status = 'break_detected'
            break_category = 'TTS formatting issue'
            break_reason = 'Agent output contained markdown or numbered lists unsuitable for voice TTS synthesis.'
            prompt_fix = 'Add strict formatting prohibition in system prompt: "Do not use markdown, bullets, or numbers."'
        elif 'in stock for $' in agent_response.lower() or 'free iphone' in agent_response.lower():
            status = 'break_detected'
            break_category = 'Hallucination'
            break_reason = 'Agent generated unverified pricing or non-catalog items.'
            prompt_fix = 'Enforce fallback instruction: "Never state prices not supplied by backend tool calls."'

        new_log = CallLog(
            id=f"call_{int(time.time()*1000)}",
            conversationId=conversation_id,
            merchantId=merchant.id,
            merchantName=merchant.name,
            customerTranscript=customer_transcript,
            agentResponse=agent_response,
            durationSeconds=max(8, round(word_count * 0.7)),
            detectedIntent=detected_intent,
            detectedLanguage=detected_language,
            latencyMs=latency_ms,
            wordCount=word_count,
            status=status,
            breakCategory=break_category,
            breakReason=break_reason,
            promptFixRecommendation=prompt_fix,
            promptVersion=prompt_version,
            timestamp=datetime.utcnow().isoformat()
        )

        self.call_logs.insert(0, new_log)
        return new_log

call_log_service = CallLogService()

File: backend/src/services/callLogService.ts
import { CallLog, Merchant } from '../types';
import { mockCallLogs } from '../data/mockCallLogs';
import { languageDetector } from './languageDetector';

export class CallLogService {
  private callLogs: CallLog[] = [...mockCallLogs];

  public getCallLogs(merchantId?: string): CallLog[] {
    if (merchantId) {
      return this.callLogs.filter(c => c.merchantId === merchantId);
    }
    return this.callLogs;
  }

  public getCallById(id: string): CallLog | undefined {
    return this.callLogs.find(c => c.id === id);
  }

  /**
   * Log a new agent conversation turn and inspect for agent break patterns
   */
  public logCall(
    conversationId: string,
    merchant: Merchant,
    customerTranscript: string,
    agentResponse: string,
    detectedIntent: string,
    latencyMs: number,
    promptVersion: string
  ): CallLog {
    const wordCount = agentResponse.split(/\s+/).filter(Boolean).length;
    const detectedLanguage = languageDetector.detectLanguage(customerTranscript);

    // Agent Break Detection Logic
    let status: 'successful' | 'break_detected' = 'successful';
    let breakCategory: CallLog['breakCategory'] = undefined;
    let breakReason: string | undefined = undefined;
    let promptFixRecommendation: string | undefined = undefined;

    // 1. Long response check
    if (wordCount > merchant.maxResponseWords + 8) {
      status = 'break_detected';
      breakCategory = 'Long response';
      breakReason = `Agent response (${wordCount} words) exceeded merchant's maximum target limit (${merchant.maxResponseWords} words).`;
      promptFixRecommendation = `Update system prompt rule: "Maximum response length: ${merchant.maxResponseWords} words. Be extremely brief."`;
    }

    // 2. TTS Formatting issue check (markdown, bullets, headers)
    else if (/[\*\#\`\[\]]/.test(agentResponse) || /^\s*\d+[\.\)]/m.test(agentResponse)) {
      status = 'break_detected';
      breakCategory = 'TTS formatting issue';
      breakReason = 'Agent output contained markdown or numbered lists unsuitable for voice TTS synthesis.';
      promptFixRecommendation = 'Add strict formatting prohibition in system prompt: "Do not use markdown, bullets, or numbers."';
    }

    // 3. Hallucination check
    else if (agentResponse.toLowerCase().includes('in stock for $') || agentResponse.toLowerCase().includes('free iphone')) {
      status = 'break_detected';
      breakCategory = 'Hallucination';
      breakReason = 'Agent generated unverified pricing or non-catalog items.';
      promptFixRecommendation = 'Enforce fallback instruction: "Never state prices not supplied by backend tool calls."';
    }

    const newLog: CallLog = {
      id: `call_${Date.now()}`,
      conversationId,
      merchantId: merchant.id,
      merchantName: merchant.name,
      customerTranscript,
      agentResponse,
      durationSeconds: Math.max(8, Math.round(wordCount * 0.7)),
      detectedIntent,
      detectedLanguage,
      latencyMs,
      wordCount,
      status,
      breakCategory,
      breakReason,
      promptFixRecommendation,
      promptVersion,
      timestamp: new Date().toISOString()
    };

    this.callLogs.unshift(newLog);
    return newLog;
  }
}

export const callLogService = new CallLogService();
